Remove IteratorRange in favor of llvm::iterator_range

jrose-apple opened the pull request:
Now that `llvm::iterator_range` has `empty`, there's not enough reason to keep our own version of it in the Swift repo.

No functionality change.

jrose-apple committed:
Remove IteratorRange in favor of llvm::iterator_range

Now that llvm::iterator_range has 'empty', there's not enough reason to
keep our own version of it in the Swift repo.

No functionality change.

jrose-apple commented:
apple/swift-llvm#174
@swift-ci Please smoke test

gottesmm commented on include/swift/Basic/STLExtras.h:
```diff
@@ -251,37 +251,13 @@ inline Iterator prev_or_begin(Iterator it, Iterator begin) {
 /// An iterator that transforms the result of an underlying bidirectional
 /// iterator with a given operation.
 ///
+/// Slightly different semantics from llvm::map_iterator, but we should
```
What are the different semantics here? Just seems out of place with the rest of the PR. Can you elaborate in the code?

jrose-apple replied:
It's something about whether the underlying iterator returns a reference or not. I don't actually remember the details, but I tried to remove it and failed and I want to leave a warning for others to come.

jrose-apple replied:
It is *not* the purpose of the PR, as you note, so if you object to it going in in this form I'll just drop it.

gottesmm commented on include/swift/AST/DeclContext.h:
```diff
@@ -667,7 +667,7 @@ class DeclIterator {
 /// The range of declarations stored within an iterable declaration
 /// context.
-typedef IteratorRange<DeclIterator> DeclRange;
+typedef llvm::iterator_range<DeclIterator> DeclRange;
```
`using`?

jrose-apple replied:
I mostly wasn't updating existing code, but sure.

gottesmm reviewed (approved):
Overall looks good to me. Two small nits and a request. We use this a lot. Why not inject this into LLVM.h and strip the llvm::?

jrose-apple commented:
> Why not inject this into LLVM.h and strip the llvm::?

Sure, will do.

jrose-apple committed:
Add llvm::iterator_range to LLVM.h

If we're going to get rid of swift::IteratorRange, let's make
llvm::iterator_range easy to use.

No functionality change.

jrose-apple commented:
apple/swift-llvm#174
@swift-ci Please smoke test

jrose-apple commented:
Hm, guess I have to wait until the LLVM change propagates into `stable`.

jrose-apple commented:
@swift-ci Please smoke test

jrose-apple commented:
@swift-ci Please smoke test macOS
